alws/utils/multilib.py

```diff
@@ -19,6 +24,252 @@
+            return self._is_multilib_needed
+
+        query = select(models.BuildTask.arch).where(sqlalchemy.and_(
+                models.BuildTask.build_id == self._build_task.build_id,
+                models.BuildTask.index == self._build_task.index,
+                models.BuildTask.status == BuildTaskStatus.COMPLETED,
+        ))
+        result = True
+        async for db in get_db():
+            db_build_tasks = await db.execute(query)
+            task_arches = list(db_build_tasks.scalars().all())
+            if 'i686' not in task_arches:
+                result = False
+            self._is_multilib_needed = result
+        return result
+
+    @staticmethod
+    async def parse_response(
+            query: str, beholder_response: dict) -> typing.List[dict]:
+        result = jmespath.search(query, beholder_response)
+        result = jmespath.search(
+             "[*].{name: name, version: version, "
+             "is_multilib: repos[?arch=='x86_64'].arch[] | "
+             "contains(@, 'x86_64')}",
+             result,
+         )
+        return result if result else []
+
+    async def call_for_packages(
+            self, platform, src_rpm: str) -> typing.List[dict]:
+        ref_name = get_clean_distr_name(platform.name)
+        ref_ver = platform.distr_version
+        endpoint = f'api/v1/distros/{ref_name}/{ref_ver}/project/{src_rpm}'
+        response = await self.__call_beholder(endpoint)
+        query = (
+            "packages[?arch=='i686']"
+            ".{name: name, version: version, repos: repositories}"
+        )
+        return await self.parse_response(query, response)
+
+    async def call_for_module_artifacts(self, platform) -> typing.List[dict]:
+        if not self._module_index:
+            return []
+
+        ref_name = get_clean_distr_name(platform.name)
+        ref_ver = platform.distr_version
+        module_name = self._build_task.rpm_module.name
+        module_stream = self._build_task.rpm_module.stream
+        endpoint = (f'api/v1/distros/{ref_name}/{ref_ver}/module/'
+                    f'{module_name}/{module_stream}/x86_64/')
+        response = await self.__call_beholder(endpoint)
+        query = (
+            "artifacts[*].packages[?arch=='i686']"
+            ".{name: name, version: version, repos: repositories}[]"
+        )
+        packages = await self.parse_response(query, response)
+        if self._module_index and self._module_index.has_devel_module():
+            module_name = f'{module_name}-devel'
+            endpoint = (f'api/v1/distros/{ref_name}/{ref_ver}/module/'
+                        f'{module_name}/{module_stream}/x86_64/')
+            response = await self.__call_beholder(endpoint)
+            devel_packages = await self.parse_response(query, response)
+            packages += devel_packages
+            # Deduplicate packages
+            packages_mapping = {f'{i["name"]}-{i["version"]}': i
+                                for i in packages}
```

Comment:
can we lose some packages if regular and devel package have same name and version?

Reply:
The packages in repo are the same. The only difference is where they will be put inside modules metadata and there everything seems to be working great